docs/root/configuration/upstream/cluster_manager/cds.rst

```diff
@@ -18,3 +18,40 @@ Statistics
 ----------
 
 CDS has a :ref:`statistics <subscription_statistics>` tree rooted at *cluster_manager.cds.*
+
+On-demand CDS
+-------------
+
+Similar to VHDS on-demand feature in terms of hosts, the on-demand CDS API is an additional API
+that Envoy will call to dynamically fetch upstream clusters which Envoy interested in spontaneously.
+
+By default in CDS, all cluster configurations are sent to every Envoy instance in the mesh. The
+delta CDS provides the ability that the xDS management server can send incremental CDS to the Envoy
+instance, but Envoy instance can not feedback upstream clusters it interested in spontaneously, in
+other words, Envoy instance only can receive the CDS passively.
+
+In order to fix this issue, on-demand CDS uses the delta xDS protocol to allow a cluster configuration
+to be subscribed to and the necessary cluster configuration to be requested as needed. Instead
+of sending all cluster configuration or cluster configuration the Envoy instance aren't interested
+in, using on-demand CDS will allow an Envoy instance to subscribe and unsubscribe from a list of
+cluster configurations stored internally in the xDS management server. The xDS management server
+will monitor the list and use it to filter the configuration sent to an individual Envoy instance
+to only contain the subscribed cluster configurations.
+
+Subscribing to resources
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+On-demand CDS allows resources to be :ref:`subscribed <xds_protocol_delta_subscribe>` to using
+a :ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+with the :ref:`type_url <envoy_api_field_DeltaDiscoveryRequest.type_url>` set to
+`type.googleapis.com/envoy.api.v2.Cluster` and
+:ref:`resource_names_subscribe <envoy_api_field_DeltaDiscoveryRequest.resource_names_subscribe>`
+set to a list of cluster resource names for which it would like configuration.
```

Comment:
Does the cluster name have any special form? How does the extension (RocketMQ in your case) know what cluster to ask for? I.e. how is it translating something in the request to something it can ask for from the management server?

Reply:
Well, no special form here. the `resource_names_subscribe` would notify management server cluster name the envoy instance need, the `type_url` would notify management server the type of resource the envoy instance need. The management server will translate the clusters into `resources` in `DeltaDiscoveryResponse` envoy can accept.

Reply:
I'm still trying to understand how we go from some RocketMQ concept to a cluster resource name. Can you elaborate on this?

Reply:
Ok, there are series concepts in RocketMQ: **topic**, **broker**, and **nameserver**. **nameserver** is the server who provides service-discovery to find **broker** through **topic**. RocketMQ client will choose a **topic** to send/consume message. **topic** is just a name or an abstract concept, it is a set of **brokers** actually. 

In a standard procedure, RocketMQ client will fetch the broker's address through **nameserver**, the **nameserver** will send the route information to the client. Client would send/consume message to broker directly. 

In service mesh, one topic's route information is abstracted into one cluster in xDS, the topic name is the cluster's name or the resource's name in delta xDS. Sometimes the envoy instance do not retain the cluster ( can not find the cluster in `clusterManager` ) but RocketMQ client needs it, the `addToClusterInterest` method will be trigger to fetch the route information from management server, and the `ClusterUpdateCallbacks` will notify the the events to be processed.

Reply:
So, I think the interesting question here is how do we map topic to cluster name? Is this always going to be 1:1? Is the RocketMQ filter config going to allow for some kind of transformation or mapping to happen?

Reply:
Yes, it's 1: 1 now.

Reply:
@htuch Yes, it's 1:1 now, I think 1:1 is the most suited way for our architecture.

Reply:
I'm wondering when other folks pick up RocketMQ whether they are going to feel the same way. Many Envoy operators have stylized ways of naming clusters that reflect things like their provenance, tenant, workload, version, etc. Would something like `cluster-foo-v145-acme-tau-beta-lambda` be a reasonable abstract topic to route to? Or are topics more like `the-weather-today-in-france` and you are expecting the clusters to match this logical naming?

Reply:
@htuch  As a developer of Apache RocketMQ, I would say 1:1 mapping RocketMQ topic to envoy cluster is the most natural way.  One topic is normally served by multiple broker servers to maintain availability and each broker serves multiple topics.  The topic name itself can be any arbitrary valid name. Prod deployment usually follows some naming schemes to implement multi-tenant etc.  From the perspective of application developer and operator, it is OK to treat a topic as a cluster of virtual shared servers.

" Would something like cluster-foo-v145-acme-tau-beta-lambda be a reasonable abstract topic to route to? Or are topics more like the-weather-today-in-france and you are expecting the clusters to match this logical naming?"

The name of the topic is up to the application developer/operators.  In most cases, topic name follows TenantId-productName-moduleX-featureY paradigm. Application developers may version their topics.